### DISCUSSION TITLE
ZKP #4 Office Hour #2 (10.29) 视频 47:50 有个疑问

### DISCUSSION THREAD

**Discussion by xrqin** [z2o-k7e/halo2-step-by-step#62] [Q&A] [answered]
这里说对于Is_Zero电路，有两种约束方法。第一种是不用inv，第二种是用inv。不用的inv的约束是val*is_zero == 0

(1-is_zero)*(1-val) == 0，用inv的约束是val*(1-val*inv) == 0。Po老师说用inv都是两行，但是少一行约束，所以第二种更efficent。我想问但是用inv的话也多了一列（装inv）啊？这个难道不是tradeoff吗？难道约束带来的improvement更多？

**Comment by pochenai** [comment, marked as answer]
你理解的没问题，是一种trade off, 第二种在很多行都需要is_zero约束的时候才会有更大的优势
